lib/memory_profiler/results.rb

```diff
@@ -47,16 +47,17 @@ def register_results(allocated, retained, top)
     end
 
     def string_report(data, top)
-      data.values
-          .keep_if { |stat| stat.string_value }
-          .map! { |stat| [stat.string_value, stat.location] }
-          .group_by { |string, _location| string }
-          .sort_by {|string, list| [-list.size, string] }
-          .first(top)
-          .map { |string, list| [string, list.group_by { |_string, location| location }
-                                             .map { |location, locations| [location, locations.size] }
-                                ]
-          }
+      data.values.
+        keep_if { |stat| stat.string_value }.
+        # Hash by string and object_id. Strings >200 chars have been shortened
+        # and will have different object_ids
+        group_by { |stat| "#{stat.string_value} - #{stat.string_value.object_id}" }.
```

Comment:
I also like that this gets rid of the `.map!` call, since the `top_n` portion of the profiling usually is a large portion of the CPU time once gathering the data has finished.  Though, the more lengthy parts are probably in the `.group_by` and `.sort_by` calls.

Reply:
So I did do some tests with this change the string to array change in place (since I also wanted to confirm what was in this PR was also far superior to what I had in #48), using the `ManageIQ` app as a test bed, and it seems like it was basically the same speed and a bit more memory efficient to use the arrays over the strings (not by a whole lot, but still).

Up to you on whether or not you think it is worth it, but here are some of the results:


**"Script" Used**

```console
$ bin/rails c
** Using session_store: ActionDispatch::Session::MemCacheStore
Loading development environment (Rails 5.0.6)
irb(main):001:0> require 'memory_profiler'; report = nil; Benchmark.measure { report = MemoryProfiler.report { _ "foo" } }
=> #<Benchmark::Tms:0x007fd498f69f40 ...>
irb(main):002:0> exit
```


**Before**

```
Run 1:  time => 68.65s, mem => 609.2MB
Run 2:  time => 67.43s, mem => 606.6MB
Run 3:  time => 69.74s, mem => 615.1MB
```


**After**

```
Run 1:  time => 68.32s, mem => 539.7MB (peak ~550MB)
Run 2:  time => 68.61s, mem => 533.7MB
Run 3:  time => 68.14s, mem => 531.6MB (peak ~538MB)
```

_**Note:**  The method to get the memory was relatively crude, since I was effectively watching a process monitor every time I started the "script", and recorded the memory results after the block had finished, rebooting the console after each run.  I avoided capturing memory within the same process right after the block to avoid skewing the results further, and potentially triggering garbage collection after the block had run._

_The values with "peak" were where I observed a higher bit of memory slightly before the process dropped down to the first value after the `.report` block had finished (what I only can assume is a `GC` run).  The first value is the "resting" memory at that point._


**Diff of code changes**

```diff
diff --git a/lib/memory_profiler/results.rb b/lib/memory_profiler/results.rb
index f01dea4..076b893 100644
--- a/lib/memory_profiler/results.rb
+++ b/lib/memory_profiler/results.rb
@@ -51,8 +51,9 @@ module MemoryProfiler
         keep_if { |stat| stat.string_value }.
         # Hash by string and object_id. Strings >200 chars have been shortened
         # and will have different object_ids
-        group_by { |stat| "#{stat.string_value} - #{stat.string_value.object_id}" }.
-        sort_by { |group_key, list| [-list.size, group_key] }.
+        group_by { |stat| [stat.string_value, stat.string_value.object_id] }.
+        sort_by { |group_key, list| [-list.size, group_key[0]] }.
         first(top).
         # Return array of [string, [location, count]]
         map! { |_group_key, list| [list[0].string_value,
```